### PR TITLE
Add LangChainClient and notebook demonstrating Running a chain on a dataset

### PR DESCRIPTION
**Pull request by vowelparrot**
Will open a different one with a squash of this later, but draft for testing out:

Further unit tests needed for the client methods. Also would appreciate preferences on things like
- Which directory to place the notebook in
- Is `langchain/client` a good place for the client
- Should I rm a lot of the async methods? (the ones that were wanted are the chain running ones so we can do multiple in parallel, but the others were so that I'd call mostly async methods from the async method, etc. Fine removing for now for many reasons)
- Naming of our CRUD verbs. Currently the API says "Read" for both the single and multiple return actions. I currently have "Read" and "list" in the client. 
- I used an asyncio Queue so we don't have to block on each batch end. There may be some issues I'm not familiar with that would make that less viable at this point.

Think it's probably best to wait to expose on the top level until later also.

### CODE REVIEW

File: langchain/callbacks/tracers/langchain.py [hunk 1]
```diff
@@ -138,6 +138,7 @@ def _convert_run(run: Union[LLMRun, ChainRun, ToolRun]) -> Run:
             session_id=run.session_id,
             run_type=run_type,
             parent_run_id=run.parent_uuid,
+            example_id=run.example_id,
```
Comment (agola11) [inline langchain/callbacks/tracers/langchain.py:141]: We don't use this now, but the proper field is `reference_example_id` on the `Run` schema

File: langchain/callbacks/manager.py [hunk 2]
```diff
@@ -44,11 +44,13 @@ def get_openai_callback() -> Generator[OpenAICallbackHandler, None, None]:
...
 @contextmanager
 def tracing_enabled(
-    session_name: str = "default",
+    session_name: str = "default", example_id: Optional[str] = None
 ) -> Generator[TracerSession, None, None]:
     """Get Tracer in a context manager."""
     cb = LangChainTracer()
     session = cb.load_session(session_name)
+    if example_id:
+        cb.example_id
```
Comment (agola11) [inline langchain/callbacks/manager.py:53]: cb.example_id = example_id

File: docs/use_cases/evaluation/evaluating_traced_examples.ipynb
Comment (agola11) [inline docs/use_cases/evaluation/evaluating_traced_examples.ipynb:1]: let's wrap this all in a client that holds the endpoint, api key, etc. this way we don't have to pass it to all these one-off methods

File: langchain/callbacks/tracers/langchain.py (hunk 1 above)
Comment (vowelparrot) [reply langchain/callbacks/tracers/langchain.py:141]: Got it so I'll change to `reference_example_id=run.example_id,`

File: langchain/client/langchain.py [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+    def list_examples(self, dataset_id: Optional[str] = None) -> Iterable[Example]:
+        """List the datasets on the LangChain+ API."""
+        params = {} if dataset_id is None else {"dataset": dataset_id}
+        response = requests.get(
+            self.api_url + "/examples", headers=self._headers, params=params
+        )
+        _raise_rich_error(response)
+        return [Example(**dataset) for dataset in response.json()]
+
+    async def alist_examples(
```
[the remark this answers is not in the document]
Comment (agola11) [reply langchain/client/langchain.py]: would use `list` for getting multiple, `read` for getting one

File: langchain/client/langchain.py [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+        response.raise_for_status()
+    except HTTPError as e:
+        raise ValueError(response.text) from e
+
+
+class LangChainPlusClient(BaseSettings):
+    """Client for interacting with the LangChain+ API."""
+
+    api_key: str = Field(default=None, env="LANGCHAIN_API_KEY")
+    api_url: str = Field(default="http://localhost:8000", env="LANGCHAIN_ENDPOINT")
```
[the remark this answers is not in the document]
Comment (agola11) [reply langchain/client/langchain.py]: default is fine

File: langchain/callbacks/manager.py [hunk 5]
```diff
@@ -66,7 +67,9 @@ def tracing_v2_enabled(
         "This is not yet stable and may change in the future."
     )
     cb = LangChainTracerV2()
-    session = cb.load_session(session_name)
+    if example_id:
+        cb.example_id = UUID(example_id)
```
Comment (nfcampos) [inline langchain/callbacks/manager.py:71]: Wouldn't this be better as an optional constructor arg to the tracer?

File: langchain/callbacks/tracers/schemas.py [hunk 6]
```diff
@@ -115,7 +119,6 @@ class Run(BaseModel):
     parent_run_id: Optional[UUID]
     reference_example_id: Optional[UUID]
     run_type: RunTypeEnum
-    child_runs: List[Run] = Field(default_factory=list)
```
Comment (nfcampos) [inline langchain/callbacks/tracers/schemas.py:118]: Isnt this still needed for the old tracer?

File: langchain/client/langchain.py [hunk 7]
```diff
@@ -0,0 +1,521 @@
...
+
+    async def arun_on_dataset(
+        self,
+        dataset_name: str,
+        llm_or_chain: Union[Chain, BaseLanguageModel],
+        num_workers: int = 5,
+        num_repetitions: int = 1,
+        session_name: Optional[str] = None,
+        verbose: bool = False,
+    ) -> Dict[str, Any]:
```
Comment (nfcampos) [inline langchain/client/langchain.py:367]: A suggestion here: I think what you're trying to achieve here is a  version of `asyncio.gather` with a concurrency limit, which can be implemented this way https://stackoverflow.com/a/61478547

File: langchain/client/langchain.py [hunk 8]
```diff
@@ -0,0 +1,521 @@
...
+            for i, example in enumerate(examples):
+                result = self.run_llm_or_chain(
+                    example,
+                    tracer,
+                    llm_or_chain,
+                    num_repetitions,
+                )
+                if verbose:
+                    print(f"{i+1} processed", flush=True, end="\r")
+                if i > 5:
```
Comment (nfcampos) [inline langchain/client/langchain.py:518]: What does 5 mean?

**Review by nfcampos — COMMENTED** [review]
Some comments

File: langchain/client/langchain.py (hunk 7 above)
Comment (vowelparrot) [reply langchain/client/langchain.py:367]: Oh cool love it .So make N semaphors then and grab?

File: langchain/client/langchain.py (hunk 8 above)
Comment (vowelparrot) [reply langchain/client/langchain.py:518]: oops sorry was debugging. I'll turn this into a draft again

File: langchain/callbacks/manager.py (hunk 5 above)
Comment (vowelparrot) [reply langchain/callbacks/manager.py:71]: It would indeed, I believe

File: langchain/callbacks/tracers/schemas.py (hunk 6 above)
Comment (vowelparrot) [reply langchain/callbacks/tracers/schemas.py:118]: I believe the `Run` object is exclusively used in the V2 tracer right now

File: langchain/client/langchain.py (hunk 7 above)
Comment (vowelparrot) [reply langchain/client/langchain.py:367]: Will try to apply

File: langchain/callbacks/tracers/schemas.py (hunk 6 above)
Comment (vowelparrot) [reply langchain/callbacks/tracers/schemas.py:118]: Oops I was wrong it's used by the n ew one as well still

**Comment by vowelparrot** [comment]
Incorporated feedback into https://github.com/hwchase17/langchain/pull/4198

**Pull request closed without merging**